NWL/NWL.py

```diff
@@ -0,0 +1,451 @@
+    theta = find_coords(vmec, wall_s, phi, pt)
+
+    return phi, theta
+
+
+def extract_coords(source_file):
+    """Extracts Cartesian coordinates of particle surface crossings given an
+    OpenMC surface source file.
+
+    Arguments:
+        source_file (str): path to OpenMC surface source file.
+    
+    Returns:
+        coords (array of array of float): Cartesian coordinates of all particle
+            surface crossings.
+    """
+    import h5py
+    
+    # Load source file
+    file = h5py.File(source_file, 'r')
+    # Extract source information
+    dataset = file['source_bank']
+    # Extract coordinates of particle crossings
+    coords = dataset['r']
+
+    return coords
+
+
+def plot(NWL_mat, phi_bins, theta_bins, num_levels):
+    """Generates contour plot of NWL.
+
+    Arguments:
+        NWL_mat (array of array of float): NWL solutions at centroids of
+            (phi, theta) bins (MW).
+        phi_bins (array of float): centroids of toroidal angle bins (rad).
+        theta_bins (array of float): centroids of poloidal angle bins (rad).
+        num_levels (int): number of contour regions.
+    """
+    import matplotlib.pyplot as plt
+    
+    phi_bins = np.rad2deg(phi_bins)
+    theta_bins = np.rad2deg(theta_bins)
+
+    levels = np.linspace(np.min(NWL_mat), np.max(NWL_mat), num = num_levels)
+    fig, ax = plt.subplots()
+    CF = ax.contourf(phi_bins, theta_bins, NWL_mat, levels = levels)
+    cbar = plt.colorbar(CF)
+    cbar.ax.set_ylabel('NWL (MW)')
+    plt.xlabel('Toroidal Angle (degrees)')
+    plt.ylabel('Poloidal Angle (degrees)')
+    fig.savefig('NWL.png')
+
+
+def NWL_plot(
+    source_file, ss_file, plas_eq, tor_ext, pol_ext, wall_s, num_phi = 101, num_theta = 101, num_levels = 10
+    ):
+    """Computes and plots NWL.
+
+    Arguments:
+        source_file (str): path to OpenMC surface source file.
+        ss_file (str): source strength input file.
+        plas_eq (str): path to plasma equilibrium NetCDF file.
+        tor_ext (float): toroidal extent of model (deg).
+        pol_ext (float): poloidal extent of model (deg).
+        wall_s (float): closed flux surface label extrapolation at wall.
+        num_phi (int): number of toroidal angle bins (defaults to 101).
+        num_theta (int): number of poloidal angle bins (defaults to 101).
+        num_levels (int): number of contour regions (defaults to 10).
+    """
+    import read_vmec
+    
+    tor_ext = np.deg2rad(tor_ext)
+    pol_ext = np.deg2rad(pol_ext)
+    
+    coords = extract_coords(source_file)
+    
+    # Load plasma equilibrium data
+    vmec = read_vmec.vmec_data(plas_eq)
+
+    phi_bins = np.linspace(0.0, tor_ext, num = num_phi)
+    theta_bins = np.linspace(-pol_ext/2, pol_ext/2, num = num_theta)
+    
+    # Initialize count matrix
+    count_mat = np.zeros((num_phi, num_theta))
+
+    for pt in coords:
+        # Extract Cartesian coordinates and format as array
+        pt = [pt['x'], pt['y'], pt['z']]
```

Comment:
I wonder if there is a way to extract this info with h5py to get them in tuples of size three directly, rather than named like this?

Reply:
I couldn't find a clean way to do it via h5py directly but I think I came up with a decent alternative. New commit coming soon